### PR TITLE
PERF: Implement round on the block level

### LINKED ISSUE #17254: PERF: DataFrame.round() unnecessarily slow copared to np.round()
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

np_df = np.random.randn(10000, 4000)
df = pd.DataFrame(np_df)

%timeit np.round(np_df, 2)
# 416 ms ± 10.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit df.round(2)
# 1.69 s ± 27.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit np.round(df, 2)
# 1.74 s ± 112 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
#### Problem description

Completely unexpected DataFrage.round() showed up as a major hotspot during profiling.
When looking at the code, we see that even when rounding the complete data frame to a given number of decimals it is split into series objects which are then rounded.

I am wondering if there is a reason not to pass the underlying data frame to numpy and do the rounding there in this case.

A quick test showed that something like this would give us the numpy performance:
```python
def faster_round(df, decimals):
    rounded = np.round(df.values, decimals)
    return pd.DataFrame(rounded, columns=df.columns, index=df.index)

%timeit faster_round(df, 2)
# 417 ms ± 14.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #17254 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Benchmarks are showing same speedup as OP.

```
       before           after         ratio
     [c133327c]       [1a2c35f5]
     <main>           <faster-round>
-            115M              72M     0.63  frame_methods.Round.peakmem_round_transposed
-      1.24±0.2ms       60.9±0.9μs     0.05  frame_methods.Round.time_round
-        863±70ms       61.2±0.4μs     0.00  frame_methods.Round.time_round_transposed

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -9932,15 +9936,19 @@ def _series_round(ser: Series, decimals: int) -> Series:
...
+            new_mgr = self._mgr.apply("round", decimals=decimals)
         else:
             raise TypeError("decimals must be an integer, a dict-like or a Series")
 
-        if len(new_cols) > 0:
+        if new_cols is not None and len(new_cols) > 0:
             return self._constructor(
                 concat(new_cols, axis=1), index=self.index, columns=self.columns
             ).__finalize__(self, method="round")
+        elif new_mgr is not None:
```
Comment (jbrockmendel) [inline pandas/core/frame.py:9948]: is new_mgr only assigned in one place above?  if so, why not just return there?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -9915,6 +9915,7 @@ def _dict_round(df: DataFrame, decimals):
                     yield _series_round(vals, decimals[col])
                 except KeyError:
                     yield vals
+            df._clear_item_cache()
```
Comment (jbrockmendel) [inline pandas/core/frame.py:9918]: why is this needed?

Comment (lithomas1) [reply pandas/core/frame.py:9918]: Oops, this snuck in. I'm going to make another PR for this separately.

File: pandas/core/frame.py (hunk 1 above)
Comment (lithomas1) [reply pandas/core/frame.py]: Thanks. updated.

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -1397,6 +1397,23 @@ def quantile(
...
+        Parameters
+        ----------
+        decimals: int,
+            Number of decimal places to round to.
+            Caller is responsible for validating this
+        using_cow: bool,
+            Whether Copy on Write is enabled right now
+        """
+        if not self.is_numeric or self.is_bool:
+            return self.copy(deep=None if using_cow else True)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1414]: possibly raise?

Comment (lithomas1) [reply pandas/core/internals/blocks.py:1414]: I believe that DataFrame.round() never raises even when there are no numeric columns, but Series.round() does.

Do you want me to add a comment that this doesn't raise?

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1414]: if this is just a perf PR than this should maintain behavior, but might merit an issue or TODO comment.  looks like some of our EAs have round methods that are not documented as part of the interface.

File: pandas/core/internals/blocks.py [hunk 4]
```diff
@@ -1397,6 +1397,32 @@ def quantile(
...
+        Parameters
+        ----------
+        decimals: int,
+            Number of decimal places to round to.
+            Caller is responsible for validating this
+        using_cow: bool,
+            Whether Copy on Write is enabled right now
+        """
+        if not self.is_numeric or self.is_bool:
+            return self.copy(deep=using_cow)
```
Comment (phofl) [inline pandas/core/internals/blocks.py:1416]: not using_cow I think?

Comment (lithomas1) [reply pandas/core/internals/blocks.py]: Thanks for the catch.

File: pandas/core/internals/blocks.py [hunk 5]
```diff
@@ -1421,6 +1421,32 @@ def quantile(
...
+            Whether Copy on Write is enabled right now
+        """
+        if not self.is_numeric or self.is_bool:
+            return self.copy(deep=not using_cow)
+        # TODO: round only defined on BaseMaskedArray
+        # Series also does this, so would need to fix both places
+        # error: Item "ExtensionArray" of "Union[ndarray[Any, Any], ExtensionArray]"
+        # has no attribute "round"
+        return new_block_2d(
+            self.values.round(decimals),  # type: ignore[union-attr]
```
Comment (phofl) [inline pandas/core/internals/blocks.py:1446]: I think you want to use ``self.make_block_same_class`` here? If class could change then ``self.make_block``. Don't have to set placement there

File: pandas/core/frame.py [hunk 6]
```diff
@@ -9934,12 +9934,16 @@ def _series_round(ser: Series, decimals: int) -> Series:
                 raise TypeError("Values in decimals must be integers")
             new_cols = list(_dict_round(self, decimals))
         elif is_integer(decimals):
-            # Dispatch to Series.round
-            new_cols = [_series_round(v, decimals) for _, v in self.items()]
+            # Dispatch to Block.round
+            return self._constructor(
+                self._mgr.round(decimals=decimals, using_cow=using_copy_on_write()),
+                index=self.index,
```
Comment (phofl) [inline pandas/core/frame.py:9940]: Are index and columns here necessary? These information are tied to the manager

File: pandas/core/internals/blocks.py (hunk 5 above)
Comment (phofl) [reply pandas/core/internals/blocks.py:1446]: This is dangerous, I don't think that round guarantees a copy, e.g. 

```
arr = np.array([1, 2])
arr2 = arr.round(2)
arr2[0] = 100
```

You can check this with

```
arr is arr2
```

(also please add a test for this case)

File: pandas/core/frame.py (hunk 6 above)
Comment (lithomas1) [reply pandas/core/frame.py]: Thanks, this slipped through the copy paste.

File: pandas/core/internals/blocks.py (hunk 5 above)
Comment (lithomas1) [reply pandas/core/internals/blocks.py:1479]: Thanks for the tip.
I found the source here https://github.com/numpy/numpy/blob/486878b37fc7439a3b2b87747f50db9b62fea8eb/numpy/core/src/multiarray/calculation.c#L625-L636
and it looks like this is only triggered for integer case with decimals >= 0.

BTW, it also looks like round is an operation that can be done ``inplace`` by using the ``out`` parameter in numpy.
Maybe I should open a PR for that as a follow-up?

Comment (phofl) [reply pandas/core/internals/blocks.py:1479]: Would have to check if it's actually faster, but if yes then should utilise it in CoW cases

Comment (phofl) [reply pandas/core/internals/blocks.py:1479]: Follow up sounds good to me

File: pandas/core/internals/blocks.py [hunk 7]
```diff
@@ -1447,6 +1447,40 @@ def quantile(
...
+            Whether Copy on Write is enabled right now
+        """
+        if not self.is_numeric or self.is_bool:
+            return self.copy(deep=not using_cow)
+        # If self.values is a numpy array of type int
+        # and decimals >= 0, this is a no-op and numpy
+        # returns the same array, so need to set refs
+        # https://github.com/numpy/numpy/blob/486878b37fc7439a3b2b87747f50db9b62fea8eb/numpy/core/src/multiarray/calculation.c#L625-L636
+        refs = None
+        if not self.is_extension and self.dtype.kind == "i" and decimals >= 0:
```
Comment (phofl) [inline pandas/core/internals/blocks.py:1472]: I'd prefer checking:

```
result = self.values.round(..)
if self.values is result:
   refs = self.refs
```

This would not depend on an implementation detail of numpy

File: pandas/tests/copy_view/test_methods.py [hunk 8]
```diff
@@ -986,20 +986,25 @@ def test_sort_values_inplace(using_copy_on_write, obj, kwargs, using_array_manag
...
     df = DataFrame({"a": [1, 2], "b": "c"})
-    df2 = df.round()
     df_orig = df.copy()
+    df2 = df.round(decimals=decimals)
 
     if using_copy_on_write:
         assert np.shares_memory(get_array(df2, "b"), get_array(df, "b"))
-        assert np.shares_memory(get_array(df2, "a"), get_array(df, "a"))
+        if decimals >= 0:
+            assert np.shares_memory(get_array(df2, "a"), get_array(df, "a"))
```
Comment (phofl) [inline pandas/tests/copy_view/test_methods.py:998]: Can you do a else: assert not and add a todo that we can optimise with out?

File: pandas/tests/copy_view/test_methods.py [hunk 9]
```diff
@@ -986,20 +986,25 @@ def test_sort_values_inplace(using_copy_on_write, obj, kwargs, using_array_manag
...
     else:
         assert not np.shares_memory(get_array(df2, "b"), get_array(df, "b"))
 
     df2.iloc[0, 1] = "d"
+    df2.iloc[0, 0] = 4
     if using_copy_on_write:
         assert not np.shares_memory(get_array(df2, "b"), get_array(df, "b"))
+        if decimals >= 0:
```
Comment (phofl) [inline pandas/tests/copy_view/test_methods.py:1006]: if think this if is not necessary? Both cases shouldn't share

File: pandas/core/internals/blocks.py [hunk 10]
```diff
@@ -1447,6 +1447,42 @@ def quantile(
...
+        if values is self.values:
+            refs = self.refs
+            if not using_cow:
+                # Normally would need to do this before, but
+                # numpy only returns same array when round operation
+                # is no-op
+                # https://github.com/numpy/numpy/blob/486878b37fc7439a3b2b87747f50db9b62fea8eb/numpy/core/src/multiarray/calculation.c#L625-L636
+                values = values.copy()
+        return self.make_block_same_class(
+            values,
```
Comment (phofl) [inline pandas/core/internals/blocks.py:1482]: This could fit into one line?

**Review by phofl — COMMENTED** [review]
small comment, otherwise lgtm

**Review by phofl — APPROVED** [review]

**Comment by jbrockmendel** [comment]
Belated: LGTM
